flink-ml-iteration/src/main/java/org/apache/flink/iteration/progresstrack/OperatorEpochWatermarkTracker.java

```diff
@@ -0,0 +1,145 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.iteration.progresstrack;
+
+import org.apache.flink.annotation.VisibleForTesting;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Tracks the epoch watermark from each input. Once the minimum epoch watermark changed, it would
+ * notify the listener.
+ */
+public class OperatorEpochWatermarkTracker {
+
+    private final OperatorEpochWatermarkTrackerListener progressTrackerListener;
+
+    private final List<InputStatus> inputStatuses;
+
+    private final LowerBoundMaintainer allInputsLowerBound;
+
+    public OperatorEpochWatermarkTracker(
+            int[] numberOfChannels, OperatorEpochWatermarkTrackerListener progressTrackerListener) {
+        checkState(numberOfChannels != null && numberOfChannels.length >= 1);
+        this.progressTrackerListener = checkNotNull(progressTrackerListener);
+
+        this.inputStatuses = new ArrayList<>(numberOfChannels.length);
+        for (int numberOfChannel : numberOfChannels) {
+            inputStatuses.add(new InputStatus(numberOfChannel));
+        }
+
+        this.allInputsLowerBound = new LowerBoundMaintainer(numberOfChannels.length);
+    }
+
+    public void onEpochWatermark(int inputIndex, String sender, int epochWatermark)
+            throws IOException {
+        InputStatus inputStatus = inputStatuses.get(inputIndex);
+        inputStatus.onUpdate(sender, epochWatermark);
+
+        if (inputStatus.getInputLowerBound() > allInputsLowerBound.getValue(inputIndex)) {
+            int oldLowerBound = allInputsLowerBound.getLowerBound();
+            allInputsLowerBound.updateValue(inputIndex, inputStatus.getInputLowerBound());
+            if (allInputsLowerBound.getLowerBound() > oldLowerBound) {
+                progressTrackerListener.onEpochWatermarkIncrement(
+                        allInputsLowerBound.getLowerBound());
+            }
+        }
+    }
+
+    @VisibleForTesting
+    int[] getNumberOfInputs() {
```

Comment:
I am not very sure but maybe we could remove this method, which just test the internal implementation.
What is in my mind is that we could test the `OperatorEpochWatermarkTrackerListener`.